From be51401d54dccf25e13177f856efc32096e0b65a Mon Sep 17 00:00:00 2001
From: Alexandre Janniaux <ajanni@videolabs.io>
Date: Tue, 29 Aug 2023 13:50:11 +0200
Subject: [PATCH 087/100] input: don't signal EOF when input is stopped

When stopping an input, the player is signalled with vlc_player_Stop(),
which set the input for destruction in the player destructor thread.

When we called input_Stop(), the input thread will continue running, but
the demux will stop and notify the EOF state. This is reported by the
modified code to the es_out which considers it as a drain request, and
it's draining the decoders and the outputs despite the request to stop.

When calling input_Stop(), we are guaranteed that input_Stopped() will
be true afterwards, thus removing the signal removes this drain, and
removing the EOF signalling makes sense if the media was not finished
and was stopped.

(cherry picked from commit 9e432fca04fb11a0d4a7f56edc02bb77655b22cb)
---
 src/input/input.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/src/input/input.c b/src/input/input.c
index 23eb8541a7..022199ea90 100644
--- a/src/input/input.c
+++ b/src/input/input.c
@@ -602,6 +602,12 @@ static void MainLoopDemux( input_thread_t *p_input, bool *pb_changed )
         UpdateGenericFromDemux( p_input );
     }
 
+    /* If the input is stopped, we don't signal anything to the es_out and
+     * ignore the previous results from the demux, and we need to stop
+     * demuxing to let the input close as fast as possible. */
+    if (input_Stopped(p_input))
+        return;
+
     if( i_ret == VLC_DEMUXER_EOF )
     {
         msg_Dbg( p_input, "EOF reached" );
